File: java/examples/slimrpc/simple/src/main/java/com/example_service/example/client/SlimrpcGroupClientMain.java
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

package com.example_service.example.client;

import io.agntcy.slim.examples.common.Common;
import com.example_service.ExampleRequest;
import com.example_service.ExampleResponse;
import com.example_service.TestSlimrpc;
import io.agntcy.slim.bindings.App;
import io.agntcy.slim.bindings.Channel;
import io.agntcy.slim.bindings.ClientConfig;
import io.agntcy.slim.bindings.MulticastStreamMessage;
import io.agntcy.slim.bindings.Name;
import io.agntcy.slim.bindings.RpcMulticastItem;
import io.agntcy.slim.bindings.Service;
import io.agntcy.slim.bindings.SlimBindings;
import io.agntcy.slim.bindings.slimrpc.MulticastClientBidiStream;
import io.agntcy.slim.bindings.slimrpc.MulticastResponseStream;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

/**
 * Multicast (group) RPC client example.
 *
 * <p>Demonstrates broadcasting RPC calls to multiple server instances using the
 * generated {@link TestSlimrpc.TestGroupClient} stub. All four streaming patterns
 * are exercised: unary-unary, unary-stream, stream-unary, and stream-stream.
 *
 * <p>Start two server instances (server1 and server2) before running this client.
 */
public final class SlimrpcGroupClientMain {

    public static void main(String[] args) throws Exception {
        String serverAddr = Common.getServerEndpoint();
        String serversStr = "server1,server2";
        for (int i = 0; i < args.length; i++) {
            if ("--server".equals(args[i]) && i + 1 < args.length) {
                serverAddr = args[i + 1];
            } else if ("--servers".equals(args[i]) && i + 1 < args.length) {
                serversStr = args[i + 1];
            }
        }

        List<Name> serverNames = new ArrayList<>();
        for (String s : serversStr.split(",")) {
            serverNames.add(new Name(Common.NAME_ORG, Common.NAME_NS, s.trim()));
        }

        SlimBindings.initializeWithDefaults();
        Service service = SlimBindings.getGlobalService();

        Name localName = new Name(Common.NAME_ORG, Common.NAME_NS, "client");
        App app = service.createAppWithSecret(localName, Common.DEFAULT_SHARED_SECRET);

        ClientConfig clientConfig = SlimBindings.newInsecureClientConfig(serverAddr);
        long connId = service.connect(clientConfig);
        app.subscribe(app.name(), connId);

        // Group channel targeting all server instances
        Channel channel = Channel.newGroupWithConnection(app, serverNames, connId);
        TestSlimrpc.TestGroupClient client = new TestSlimrpc.TestGroupClientImpl(channel);

        try {
            System.out.println("SLIM_RPC_GROUP_CLIENT_STARTED");
            runMulticastUnary(client);
            runMulticastUnaryStream(client);
            runMulticastStreamUnary(client);
            runMulticastStreamStream(client);
            System.out.println("SLIM_RPC_GROUP_CLIENT_DONE");
        } catch (Exception e) {
            System.err.println("Group client error: " + e.getMessage());
            throw e;
        } finally {
            channel.closeAsync(null);
        }
    }

    // ---- Multicast Unary-Unary ----

    private static void runMulticastUnary(TestSlimrpc.TestGroupClient client) throws Exception {
        System.out.println("=== Multicast Unary-Unary ===");
        ExampleRequest request = ExampleRequest.newBuilder()
                .setExampleString("hello")
                .setExampleInteger(1)
                .build();

        MulticastResponseStream<ExampleResponse> stream =
                client.ExampleUnaryUnary(request, Duration.ofSeconds(5), null);

        while (true) {
            MulticastStreamMessage msg = stream.next();
            if (msg instanceof MulticastStreamMessage.End) break;
            if (msg instanceof MulticastStreamMessage.Error err) {
                System.out.println("  Error: " + err.error());
                break;
            }
            if (msg instanceof MulticastStreamMessage.Data data) {
                RpcMulticastItem item = data.item();
                ExampleResponse resp = ExampleResponse.parseFrom(item.message());
                System.out.println("  [" + item.context().source().toString() + "] " + resp);
            }
        }
    }

    // ---- Multicast Unary-Stream ----

    private static void runMulticastUnaryStream(TestSlimrpc.TestGroupClient client) throws Exception {
        System.out.println("\n=== Multicast Unary-Stream ===");
        ExampleRequest request = ExampleRequest.newBuilder()
                .setExampleString("hello")
                .setExampleInteger(1)
                .build();

        MulticastResponseStream<ExampleResponse> stream =
                client.ExampleUnaryStream(request, Duration.ofSeconds(5), null);

        while (true) {
            MulticastStreamMessage msg = stream.next();
            if (msg instanceof MulticastStreamMessage.End) break;
            if (msg instanceof MulticastStreamMessage.Error err) {
                System.out.println("  Error: " + err.error());
                break;
            }
            if (msg instanceof MulticastStreamMessage.Data data) {
                RpcMulticastItem item = data.item();
                ExampleResponse resp = ExampleResponse.parseFrom(item.message());
                System.out.println("  [" + item.context().source().toString() + "] " + resp);
            }
        }
    }

    // ---- Multicast Stream-Unary ----

    private static void runMulticastStreamUnary(TestSlimrpc.TestGroupClient client) throws Exception {
        System.out.println("\n=== Multicast Stream-Unary ===");

        MulticastClientBidiStream<ExampleRequest, ExampleResponse> stream =
                client.ExampleStreamUnary(Duration.ofSeconds(5), null);

        // Send requests
        for (int i = 0; i < 3; i++) {
            ExampleRequest req = ExampleRequest.newBuilder()
                    .setExampleString("item " + i)
                    .setExampleInteger(i)
                    .build();
            stream.send(req);
        }
        stream.closeSend();

        // Receive multicast responses
        while (true) {
            MulticastStreamMessage msg = stream.recv();
            if (msg instanceof MulticastStreamMessage.End) break;
            if (msg instanceof MulticastStreamMessage.Error err) {
                System.out.println("  Error: " + err.error());
                break;
            }
            if (msg instanceof MulticastStreamMessage.Data data) {
                RpcMulticastItem item = data.item();
                ExampleResponse resp = ExampleResponse.parseFrom(item.message());
                System.out.println("  [" + item.context().source().toString() + "] " + resp);
            }
        }
    }

    // ---- Multicast Stream-Stream ----

    private static void runMulticastStreamStream(TestSlimrpc.TestGroupClient client) throws Exception {
        System.out.println("\n=== Multicast Stream-Stream ===");

        MulticastClientBidiStream<ExampleRequest, ExampleResponse> stream =
                client.ExampleStreamStream(Duration.ofSeconds(5), null);

        // Send requests in a background thread
        Thread sendThread = new Thread(() -> {
            try {
                for (int i = 0; i < 3; i++) {
                    ExampleRequest req = ExampleRequest.newBuilder()
                            .setExampleString("item " + i)
                            .setExampleInteger(i)
                            .build();
                    stream.send(req);
                }
                stream.closeSend();
            } catch (Exception e) {
                System.out.println("Send error: " + e.getMessage());
            }
        });
        sendThread.start();

        // Receive multicast responses
        while (true) {
            MulticastStreamMessage msg = stream.recv();
            if (msg instanceof MulticastStreamMessage.End) break;
            if (msg instanceof MulticastStreamMessage.Error err) {
                System.out.println("  Error: " + err.error());
                break;
            }
            if (msg instanceof MulticastStreamMessage.Data data) {
                RpcMulticastItem item = data.item();
                ExampleResponse resp = ExampleResponse.parseFrom(item.message());
                System.out.println("  [" + item.context().source().toString() + "] " + resp);
            }
        }

        sendThread.join();
    }
}

File: kotlin/examples/slimrpc/simple/src/main/kotlin/com/example_service/example/client/SlimrpcGroupClientMain.kt
// Copyright AGNTCY Contributors (https://github.com/agntcy)
// SPDX-License-Identifier: Apache-2.0

package com.example_service.example.client

import com.example_service.ExampleResponse
import com.example_service.TestSlimrpc
import com.example_service.exampleRequest
import io.agntcy.slim.bindings.*
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.time.Duration

private const val NAME_ORG = "agntcy"
private const val NAME_NS = "grpc"
private const val DEFAULT_SERVER_ENDPOINT = "http://localhost:46357"
private const val DEFAULT_SHARED_SECRET = "my_shared_secret_for_testing_purposes_only"

fun main(args: Array<String>) = runBlocking {
    var serverAddr = System.getenv("SLIM_ADDR")?.takeIf { it.isNotEmpty() } ?: DEFAULT_SERVER_ENDPOINT
    var serversStr = "server1,server2"
    var i = 0
    while (i < args.size) {
        when (args[i]) {
            "--server" -> if (i + 1 < args.size) serverAddr = args[++i]
            "--servers" -> if (i + 1 < args.size) serversStr = args[++i]
        }
        i++
    }

    val serverNames = serversStr.split(",").map { Name(NAME_ORG, NAME_NS, it.trim()) }

    initializeWithDefaults()
    val service = getGlobalService()

    val localName = Name(NAME_ORG, NAME_NS, "client")
    val app = service.createAppWithSecret(localName, DEFAULT_SHARED_SECRET)

    val clientConfig = newInsecureClientConfig(serverAddr)
    val connId = service.connectAsync(clientConfig)
    app.subscribeAsync(app.name(), connId)

    val channel = Channel.newGroupWithConnection(app, serverNames, connId)
    val client = TestSlimrpc.TestGroupClientImpl(channel)

    try {
        println("SLIM_RPC_GROUP_CLIENT_STARTED")
        runMulticastUnary(client)
        runMulticastUnaryStream(client)
        runMulticastStreamUnary(client)
        runMulticastStreamStream(client)
        println("SLIM_RPC_GROUP_CLIENT_DONE")
    } catch (e: Exception) {
        System.err.println("Group client error: ${e.message}")
        throw e
    } finally {
        channel.closeAsync(null)
    }
}

private suspend fun runMulticastUnary(client: TestSlimrpc.TestGroupClient) {
    println("=== Multicast Unary-Unary ===")
    val request = exampleRequest {
        exampleString = "hello"
        exampleInteger = 1
    }

    val stream = client.ExampleUnaryUnary(request, Duration.ofSeconds(5), null)
    while (true) {
        val msg = stream.next()
        when (msg) {
            is MulticastStreamMessage.End -> break
            is MulticastStreamMessage.Error -> {
                println("  Error: ${msg.error}")
                break
            }
            is MulticastStreamMessage.Data -> {
                val item = msg.item
                val resp = ExampleResponse.parseFrom(item.message)
                println("  [${item.context.source}] $resp")
            }
        }
    }
}

private suspend fun runMulticastUnaryStream(client: TestSlimrpc.TestGroupClient) {
    println("\n=== Multicast Unary-Stream ===")
    val request = exampleRequest {
        exampleString = "hello"
        exampleInteger = 1
    }

    val stream = client.ExampleUnaryStream(request, Duration.ofSeconds(5), null)
    while (true) {
        val msg = stream.next()
        when (msg) {
            is MulticastStreamMessage.End -> break
            is MulticastStreamMessage.Error -> {
                println("  Error: ${msg.error}")
                break
            }
            is MulticastStreamMessage.Data -> {
                val item = msg.item
                val resp = ExampleResponse.parseFrom(item.message)
                println("  [${item.context.source}] $resp")
            }
        }
    }
}

private suspend fun runMulticastStreamUnary(client: TestSlimrpc.TestGroupClient) {
    println("\n=== Multicast Stream-Unary ===")

    val stream = client.ExampleStreamUnary(Duration.ofSeconds(5), null)
    for (i in 0 until 3) {
        val req = exampleRequest {
            exampleString = "item $i"
            exampleInteger = i.toLong()
        }
        stream.send(req)
    }
    stream.closeSend()

    while (true) {
        val msg = stream.recv()
        when (msg) {
            is MulticastStreamMessage.End -> break
            is MulticastStreamMessage.Error -> {
                println("  Error: ${msg.error}")
                break
            }
            is MulticastStreamMessage.Data -> {
                val item = msg.item
                val resp = ExampleResponse.parseFrom(item.message)
                println("  [${item.context.source}] $resp")
            }
        }
    }
}

private suspend fun runMulticastStreamStream(client: TestSlimrpc.TestGroupClient) {
    println("\n=== Multicast Stream-Stream ===")

    val stream = client.ExampleStreamStream(Duration.ofSeconds(5), null)

    coroutineScope {
        launch {
            for (i in 0 until 3) {
                val req = exampleRequest {
                    exampleString = "item $i"
                    exampleInteger = i.toLong()
                }
                stream.send(req)
            }
            stream.closeSend()
        }
    }

    while (true) {
        val msg = stream.recv()
        when (msg) {
            is MulticastStreamMessage.End -> break
            is MulticastStreamMessage.Error -> {
                println("  Error: ${msg.error}")
                break
            }
            is MulticastStreamMessage.Data -> {
                val item = msg.item
                val resp = ExampleResponse.parseFrom(item.message)
                println("  [${item.context.source}] $resp")
            }
        }
    }
}
